sheets/api.py

```diff
@@ -253,31 +261,111 @@ def batch_update_sheet_cells(self, sheet_id, request_objects):
+            be included in the header of every request Google sends to the app.
+        expiration (datetime.datetime or None): The datetime that this file watch should expire.
+            Defaults to 1 hour, and cannot exceed 24 hours.
+        credentials (google.oauth2.credentials.Credentials or None): Credentials to be used by the
+            Google Drive client
+
+    Returns:
+        dict: The Google file watch API response
+    """
+    credentials = credentials or get_credentials()
+    # Including cache_discovery=False to suppress a noisy warning
+    drive_service = build("drive", "v3", credentials=credentials, cache_discovery=False)
+    extra_body_params = {}
+    if expiration:
+        extra_body_params["expiration"] = format_datetime_for_google_timestamp(
+            expiration
+        )
+    return (
+        drive_service.files()
+        .watch(
+            fileId=file_id,
+            supportsTeamDrives=True,
+            body={
+                "id": channel_id,
+                "resourceId": file_id,
+                "address": urljoin(
+                    settings.SITE_BASE_URL,
+                    reverse("handle-coupon-request-sheet-update"),
+                ),
+                "payload": True,
+                "kind": GOOGLE_API_FILE_WATCH_KIND,
+                "type": GOOGLE_API_NOTIFICATION_TYPE,
+                "expiration": "failfailfail"
+                # **extra_body_params,
+            },
         )
+        .execute()
+    )
+
+
+def renew_coupon_request_file_watch():
+    """
+    Creates or renews a file watch on the coupon request spreadsheet depending on the existence
+    of other file watches and when they expire.
+
+    Returns:
+        (GoogleFileWatch, bool): The GoogleFileWatch object paired with a flag indicating
+            whether or not it was newly created during execution.
+    """
+    now = now_in_utc()
+    min_fresh_expiration_date = now + datetime.timedelta(
+        minutes=settings.DRIVE_WEBHOOK_RENEWAL_PERIOD_MINUTES
+    )
+    file_watches = GoogleFileWatch.objects.filter(
+        file_id=settings.COUPON_REQUEST_SHEET_ID
+    ).order_by("-expiration_date")
+    latest_watch = file_watches.first()
+    if latest_watch and latest_watch.expiration_date > min_fresh_expiration_date:
+        return latest_watch, False
+    if latest_watch.expiration_date < now:
+        log.error(
+            "Current file watch in the database is expired. Some file changes may have failed to "
+            "trigger a push notification (%s)" % latest_watch
+        )
+    expiration = now + datetime.timedelta(
+        minutes=settings.DRIVE_WEBHOOK_EXPIRATION_MINUTES
+    )
+    new_channel_id = "{}-{}".format(
+        settings.DRIVE_WEBHOOK_CHANNEL_ID, now.strftime("%Y%m%d-%H%M%S")
+    )
+    with transaction.atomic():
+        version = 1 if not latest_watch else latest_watch.version + 1
+        new_file_watch, created = GoogleFileWatch.objects.select_for_update().get_or_create(
```

Comment:
what do you think of updating the existing `GoogleFileWatch` object with fresh activation/expiration dates instead of creating a new object?  Seems like a new object will always be created here (`version = 1 if not latest_watch else latest_watch.version + 1` => `get_or_create(version=version)`).

Reply:
Good point, maintaining a single `GoogleFileWatch` object might be best here. I'll try it